File: src/java/4_BehavioralPatterns/1_ChainOfResponsibility/BrokerChain.java
package com.activemesa.behavioral.chainofresponsibility.broker;

import java.util.HashMap;
import java.util.Map;
import java.util.function.*;

class Event<TArgs>
{
//  private List<Consumer<TArgs>> handlers
//    = new ArrayList<>();
  private int index = 0;
  private Map<Integer, Consumer<TArgs>>
    handlers = new HashMap<>();

  public int subscribe(Consumer<TArgs> handler)
  {
    //handlers.add(handler);
    int i = index;
    handlers.put(index++, handler);
    return i;
  }

  public void unsubscribe(int key)
  {
    handlers.remove(key);
  }

  public void fire(TArgs args)
  {
    for (Consumer<TArgs> handler : handlers.values())
      handler.accept(args);
  }
}

class Query
{
  public String creatureName;
  enum Argument
  {
    ATTACK, DEFENSE
  }
  public Argument argument;
  public int result;

  public Query(String creatureName, Argument argument, int result)
  {
    this.creatureName = creatureName;
    this.argument = argument;
    this.result = result;
  }
}

class Game
{
  public Event<Query> queries = new Event<>();
}

class Creature
{
  private Game game;
  public String name;
  private int baseAttack, baseDefense;

  public Creature(Game game, String name,
                  int baseAttack, int baseDefense)
  {
    this.game = game;
    this.baseAttack = baseAttack;
    this.baseDefense = baseDefense;
    this.name = name;
  }

  int getAttack() {
    Query q = new Query(name, Query.Argument.ATTACK, baseAttack);
    game.queries.fire(q);
    return q.result;
  }

  int getDefense() {
    Query q = new Query(name, Query.Argument.DEFENSE, baseDefense);
    game.queries.fire(q);
    return q.result;
  }

  @Override
  public String toString() // avoid printing Game
  {
    return "Creature{" +
      "name='" + name + '\'' +
      ", attack=" + getAttack() + // !!!
      ", defense=" + getDefense() +
      '}';
  }
}

class CreatureModifier
{ // protected, not private!
  protected Game game;
  protected Creature creature;

  public CreatureModifier(Game game, Creature creature)
  {
    this.game = game;
    this.creature = creature;
  }
}

class IncreasedDefenseModifier
  extends CreatureModifier
{

  public IncreasedDefenseModifier(Game game, Creature creature)
  {
    super(game, creature);

    game.queries.subscribe(q -> {
      if (q.creatureName.equals(creature.name)
        && q.argument == Query.Argument.DEFENSE)
      {
        q.result += 3;
      }
    });
  }
}

class DoubleAttackModifier
  extends CreatureModifier
  implements AutoCloseable
{
  private int token;
  public DoubleAttackModifier(Game game, Creature creature)
  {
    super(game, creature);

    token = game.queries.subscribe(q -> {
      if (q.creatureName.equals(creature.name)
        && q.argument == Query.Argument.ATTACK)
      {
        q.result *= 2;
      }
    });
  }

  @Override // commented out exception
  public void close() /*throws Exception*/
  {
    game.queries.unsubscribe(token);
  }
}

class BrokerChainDemo
{
  public static void main(String[] args)
  {
    Game game = new Game();
    Creature goblin = new Creature(game,
      "Strong Goblin", 2, 2);

    System.out.println(goblin);

    // modifiers can be piled up
    IncreasedDefenseModifier icm = new IncreasedDefenseModifier(game, goblin);

    try (DoubleAttackModifier dam
      = new DoubleAttackModifier(game, goblin))
    {
      System.out.println(goblin);
    }

    System.out.println(goblin);
  }
}

File: src/js/4_BehavioralPatterns/1_ChainOfResponsibility/broker_chain.js
class Event
{
  constructor()
  {
    this.handlers = new Map();
    this.count = 0;
  }

  subscribe(handler)
  {
    this.handlers.set(++this.count, handler);
    return this.count;
  }

  unsubscribe(idx)
  {
    this.handlers.delete(idx);
  }

  fire(sender, args)
  {
    this.handlers.forEach(function (v, k)
    {
      v(sender, args);
    });
  }
}

let WhatToQuery = Object.freeze({
  'attack': 1,
  'defense': 2
});

class Query
{
  constructor(creatureName, whatToQuery, value)
  {
    this.creatureName = creatureName;
    this.whatToQuery = whatToQuery;
    this.value = value;
  }
}

class Game
{
  constructor()
  {
    this.queries = new Event();
  }

  performQuery(sender, query)
  {
    this.queries.fire(sender, query);
  }
}

class Creature
{
  constructor(game, name, attack, defense)
  {
    this.game = game;
    this.name = name;
    this.initial_attack = attack;
    this.initial_defense = defense;
  }

  get attack()
  {
    let q = new Query(this.name, WhatToQuery.attack,
      this.initial_attack);
    this.game.performQuery(this, q);
    return q.value;
  }

  get defense()
  {
    let q = new Query(this.name, WhatToQuery.defense,
      this.initial_defense);
    this.game.performQuery(this, q);
    return q.value;
  }

  toString()
  {
    return `${this.name}: (${this.attack}/${this.defense})`;
  }
}

class CreatureModifier
{
  constructor(game, creature)
  {
    this.game = game;
    this.creature = creature;
    this.token = game.queries.subscribe(
      this.handle.bind(this)
    );
  }

  handle(sender, query)
  {
    // implement in inheritors
  }

  dispose()
  {
    game.queries.unsubscribe(this.token);
  }
}

class DoubleAttackModifier extends CreatureModifier
{
  constructor(game, creature)
  {
    super(game, creature);
  }

  handle(sender, query) {
    if (query.creatureName === this.creature.name &&
        query.whatToQuery === WhatToQuery.attack)
    {
      query.value *= 2;
    }
  }
}

class IncreaseDefenseModifier extends CreatureModifier
{
  constructor(game, creature)
  {
    super(game, creature);
  }

  handle(sender, query)
  {
    if (query.creatureName === this.creature.name &&
        query.whatToQuery === WhatToQuery.defense)
    {
      query.value += 2;
    }
  }
}

let game = new Game();
let goblin = new Creature(game, 'Strong Goblin', 2, 2);
console.log(goblin.toString());

let dam = new DoubleAttackModifier(game, goblin);
console.log(goblin.toString());

let idm = new IncreaseDefenseModifier(game, goblin);
console.log(goblin.toString());
idm.dispose();

dam.dispose();
console.log(goblin.toString());